File: 557.Reverse-Words-in-a-String-III.go
// https://leetcode.com/problems/reverse-words-in-a-string-iii/
//
// algorithms
// Easy (62.57%)
// Total Accepted:    105,822
// Total Submissions: 169,122
// beats 100.0% of golang submissions

package leetcode

import "bytes"

func reverseWords(s string) string {
	length := len(s)

	if length == 0 {
		return s
	}

	res := bytes.NewBufferString("")
	prevIdx := 0
	for idx, n := range s {
		if n == ' ' {
			res.WriteString(reverseStr(s[prevIdx:idx]))
			res.WriteString(" ")
			prevIdx = idx + 1
		}
	}

	res.WriteString(reverseStr(s[prevIdx:]))

	return res.String()
}

func reverseStr(s string) string {
	res := []byte(s)

	head, tail := 0, len(res)-1
	for head < tail {
		res[head], res[tail] = res[tail], res[head]
		head++
		tail--
	}

	return string(res)
}

File: 557.Reverse-Words-in-a-String-III.java
// https://leetcode.com/problems/reverse-words-in-a-string-iii/
//
// algorithms
// Easy (64.06%)
// Total Accepted:    125,237
// Total Submissions: 195,508
// beats 100.0% of java submissions


class Solution {
    public String reverseWords(String s) {
        char[] chs = s.toCharArray();
        int pre = 0, i = 0, len = chs.length;

        for (; i < chs.length; i++) {
            if (chs[i] == ' ') {
                reverseCharArray(chs, pre, i - 1);
                pre = i + 1;
            }
        }

        reverseCharArray(chs, pre, len - 1);

        return new String(chs);
    }

    public void reverseCharArray(char[] chs, int i, int j) {
        while (i < j) {
            swap(chs, i, j);
            i++;
            j--;
        }
    }

    public void swap(char[] chs, int i, int j) {
        char tmp = chs[i];
        chs[i] = chs[j];
        chs[j] = tmp;
    }
}